### DISCUSSION TITLE
Level of customization possible with nemo actions

### DISCUSSION THREAD

**Discussion by anaximeno** [linuxmint/discussions#231] [Development] [answered]
The new spices module, actions for nemo, will allow customization of nemo to what level?

**Comment by rcalixte** [comment, marked as answer]
Nemo already has support for Actions. The new settings module and repository hopefully centralize the existing collections and enable greater discoverability. The scope of Actions are the context menus within Nemo, including the desktop. In terms of functionality, there are limitless possibilities via helper scripts and one-liners in the Exec key. Unlike Applets, Desklets, and Extensions, there are no obvious settings. Take a look through the Actions repository and Spices website for examples.

Some resources:
* [Sample Action file](https://github.com/linuxmint/nemo/blob/master/files/usr/share/nemo/actions/sample.nemo_action)
* @smurphos [Collection of Actions](https://github.com/smurphos/nemo_actions_and_cinnamon_scripts)

**Reply by anaximeno** [reply]
Oh, nice, could it be possible in the future for we to increase the level of extensibility for more than only the context menu? I believe we would've to adapt nemo for that, allowing others sections like the sidebar, top bar, etc, to also be extensible.

**Reply by rcalixte** [reply]
Right. It would take some reasonable development effort to make some of those changes but also, what would be the gain? There has been considerable demand for what amount to Actions in the Mint forums, Reddit, GitHub discussions, etc. What changes would you propose to other sections within Nemo? Maybe some additional settings?

**Reply by anaximeno** [reply]
Yes, it isn't an easy thing to implement specially depending on how nemo is structured.

Implementing this would certainly have many gains regarding user customization and the distribution of responsibility/effort in developing new features to the ones who created or maintain these extensions. But we could say that these future gains don't  justify a current effort in implementing this, at this moment, in relation to other demands, and I agree, hence why I suggest this just as a future update. And when I mentioned a future implemention I'm not implying the next release of Mint, or the one after it, but when it's possible and seems more valuable, and I would like to contribute to it's implemention as well. 

Regarding the demands we have to implement this, many people suggest some new features for nemo, and some are not taken into consideration or implemented because it would ony increase the complexity of the code, so, this could be a solution to distribute the complexity allowing new features to be implemented.

> What changes would you propose to other sections within Nemo? Maybe some additional settings?

I've been thinking on some nice features from other file managers I would like to see in nemo, but I'll analyze the viability of implementing them and maybe let the suggestions as an idea here in the discussion section.

Anyway, the new spice, actions, is a cool new feature, great job on implementing it!

**Reply by mtwebster** [reply]
There is an extension system, separate from actions, already available in Nemo - you can add property tabs, additional columns in the list view, menu items and location bars. We could add more functionality to this as well, or we can add things directly into Nemo if it makes sense to.

We maintain quite a few: https://github.com/linuxmint/nemo-extensions


Actions, on the other hand, only really make sense in a context menu, where you select some files, and perform some action on them. There's not really room for them to be more than that.

**Reply by anaximeno** [reply]
Oh, yeah, there's the nemo extensions repository, it would be more difficult to integrate it as spices work I guess.

**Reply by anaximeno** [reply]
@rcalixte is it possible to define an icon (symbolic) for the action? I've been testing and none of them had icons I believe.

**Reply by rcalixte** [reply]
Ah, that's an oversight on my part. I have icons in menus disabled in my settings so I hadn't noticed. I'll push out some icons shortly. Thanks for the review!

**Reply by anaximeno** [reply]
@rcalixte there's another problem. When you update the action through mintupdate, the action is being removed, and by looking at the directory `~/.local/share/nemo/actions/` the files (of the new updated version) are decompressed in the root directory for the actions (`~/.local/share/nemo/actions/`) instead of their respective directories, so some loose files can be found there (icon.png, metadata.json, etc) after an update.

**Reply by rcalixte** [reply]
Are you running Cinnamon 6.0? There are updates to the scripts to support actions.

**Reply by anaximeno** [reply]
> Are you running Cinnamon 6.0? There are updates to the scripts to support actions.

Yes, I'm alpha testing on a VM. Dowloading actions from the action module works great, but when the action has to update (you can force that too by altering the version and the last edited attrs of the metadata), I am having the problem above.

**Reply by anaximeno** [reply]
I've tested for **applets** and had no problems (didn't check for **extensions** and **desklets** but I guess they are ok too)

**Reply by rcalixte** [reply]
Nice. I don't have an alpha ISO (yet?) but based on what you've said, it's an issue with mintupdate. I don't see anything obvious from reviewing the code. I'll try to get another pair of eyes on it. (If you see what the bug is too, that would be fantastic.) Thanks for letting us know!

**Reply by anaximeno** [reply]
I don't have an alpha ISO either 😅, just using romeo and updating the source to the next version "virginia", I guess (I've turned my computer off by now, so I can't confirm the codename). But, I'll try to check what might be the problem later (it's a bit late here right now).

**Reply by anaximeno** [reply]
@rcalixte https://github.com/linuxmint/cinnamon/pull/11965

**Reply by rcalixte** [reply]
I see now. That was corrected in [Spices.py](https://github.com/linuxmint/cinnamon/blob/51de1980664ad3b522266e1db5736926f34705e5/files/usr/share/cinnamon/cinnamon-settings/bin/Spices.py#L663-L664), just not in harvester.py. That would be an oversight on my part. Apologies.
